proxy/src/telemetry/metrics/prometheus.rs

```diff
@@ -0,0 +1,487 @@
+
+            Event::StreamResponseFail(ref res, ref fail) => {
+                // TODO: do we care about the failure's error code here?
+                let labels = Arc::new(ResponseLabels::new(res, None));
+                self.update(|metrics| {
+                    *metrics.response_total(&labels) += 1;
+                    *metrics.response_duration(&labels) += fail.since_response_open;
+                    *metrics.response_latency(&labels) += fail.since_request_open;
+                });
+            },
+
+            Event::TransportOpen(_) | Event::TransportClose(_, _) => {
+                // TODO: we don't collect any metrics around transport events.
+            },
+        };
+    }
+}
+
+
+// ===== impl Serve =====
+
+impl Serve {
+    fn new(metrics: &Arc<Mutex<Metrics>>) -> Self {
+        Serve { metrics: metrics.clone() }
+    }
+}
+
+impl HyperService for Serve {
+    type Request = HyperRequest;
+    type Response = HyperResponse;
+    type Error = hyper::Error;
+    type Future = FutureResult<Self::Response, Self::Error>;
+
+    fn call(&self, req: Self::Request) -> Self::Future {
+        if req.path() != "/metrics" {
+            return future::ok(HyperResponse::new()
+                .with_status(StatusCode::NotFound));
+        }
+
+        let body = {
+            let metrics = self.metrics.lock()
+                .expect("metrics lock poisoned");
+            format!("{}", *metrics)
+        };
+        future::ok(HyperResponse::new()
+            .with_header(ContentLength(body.len() as u64))
+            .with_header(ContentType::plaintext())
+            .with_body(body))
+    }
+}
+
+
+// ===== impl RequestLabels =====
+
+impl<'a> RequestLabels {
+    fn new(req: &ctx::http::Request) -> Self {
+        let outbound_labels = if req.server.proxy.is_outbound() {
+            Some(OutboundLabels {
+                // TODO: when #429 is done, add appropriate destination label.
+                ..Default::default()
+            })
+        } else {
+            None
+        };
+
+        let authority = req.uri
+            .authority_part()
+            .map(http::uri::Authority::to_string)
+            .unwrap_or_else(String::new);
+
+        RequestLabels {
+            outbound_labels,
+            authority,
+            ..Default::default()
+        }
+    }
+}
+
+impl fmt::Display for RequestLabels {
+
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "authority=\"{}\",", self.authority)?;
+        if let Some(ref outbound) = self.outbound_labels {
+            write!(f, "direction=\"outbound\",{}", outbound)?;
+        } else {
+            write!(f, "direction=\"inbound\"")?;
+        }
+        // TODO: when PR #448 lands, use CONDUIT_PROMETHEUS_LABELS
+        // to add the appropriate labels for the owner of this pod.
+
+        Ok(())
+    }
+
+}
+
+impl fmt::Display for OutboundLabels {
+
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
```

Comment:
This is clearer!

A small thing, not required to fix: replacing the `..`s with the expected `None`s makes the match exhaustive, so any new labels will require this to be updated too.